src/foam/core/Action.js

```diff
@@ -134,59 +134,58 @@ foam.CLASS({
   ],
 
   methods: [
-    function andSlots(a, b) {
-      return this.ExpressionSlot.create({
-        args: [ a, b ],
-        code: function(a, b) {
-          return a && b;
-        }
-      });
-    },
 
-    function andSlotAndPromise(slot, promise) {
-      return this.andSlots(slot, this.PromiseSlot.create({
-        promise: promise
-      }));
+    function hasPermissions_(x, property) {
+      if ( ! this.permissionRequired || ! x.auth ) return Promise.resolve(true);
+      return Promise.all(this.permissionConfig[property].map(p => {
+        return x.auth.check(null, p);
+      })).then(arr => arr.every(b => b));
     },
 
-    function checkPermission(x, permission) {
-      if ( ! this.permissionConfig || ! x.auth ) return Promise.resolve(true);
-      return x.auth.check(null, permission);
-    },
-
-    function isEnabledFor(data) {
-      return this.isEnabled
-        ? foam.Function.withArgs(this.isEnabled, data)
-        : true;
-    },
-
-    function createIsEnabled$(data$) {
-      return this.createSlot_(this.isEnabled, 'enabled', data$);
+    function createPermissionlessCheckFor_(data$, property) {
+      if ( ! this[property] ) return foam.core.ConstantSlot.create({value: true});
+      return foam.core.ExpressionSlot.create({
+        obj$: data$,
+        code: this[property]
+      });
     },
 
-    function isAvailableFor(data) {
-      return this.isAvailable
-        ? foam.Function.withArgs(this.isAvailable, data)
-        : true;
+    function createIsEnabled$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
+            promise: this.hasPermissions_(x, 'enabled')
+          }),
+          this.createPermissionlessCheckFor_(data$, 'isEnabled')
+        ]
+      }).map(arr => arr.every(b => b));
     },
 
-    function createIsAvailable$(data$) {
-      return this.createSlot_(this.isAvailable, 'available', data$);
+    function createIsAvailable$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
+            promise: this.hasPermissions_(x, 'available')
+          }),
+          this.createPermissionlessCheckFor_(data$, 'isAvailable')
+        ]
+      }).map(arr => arr.every(b => b));
     },
 
-    async function maybeCall(ctx, data) {
+    function maybeCall(ctx, data) {
       var data$ = this.ConstantSlot.create({ value: data });
-      if (
-        this.isEnabledFor(data) &&
-        this.isAvailableFor(data) &&
-        await this.hasPermissions_('enabled', data$) &&
-        await this.hasPermissions_('available', data$)
-      ) {
+      Promise.all([
+        Promise.resolve(this.createPermissionlessCheckFor_(data$, 'isEnabled').get()),
```

Comment:
Do you need to have Promise.resolve()? Doesn't Promise.all() also accept non-promises?